File: libs/power-log-parser/src/lib/parsers/quest-completed-parser.ts
import { BlockType, CardType, GameTag } from '@firestone-hs/reference-data';
import { ParserGameTag } from '../enums';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Action, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class QuestCompletedParser implements ActionParser {
	readonly ParserName = 'QuestCompletedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		if (stateType !== StateType.PowerTaskList || node.Type !== NodeType.Action) {
			return false;
		}
		const action = node.Object as Action;
		return (
			action.Type === (BlockType.TRIGGER as number) &&
			(action.TriggerKeyword === (GameTag.SIDE_QUEST as number) ||
				action.TriggerKeyword === (GameTag.QUEST as number))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const entity = this.GameState.CurrentEntities.get(action.Entity)!;
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		if (
			this.GameState.CurrentEntities.get(action.Entity)!.GetTag(GameTag.CARDTYPE) !==
			(CardType.ENCHANTMENT as number)
		) {
			entity.SetTag(ParserGameTag.SECRET_HAS_TRIGGERED, 1);
			return [
				GameEventProvider.Create(
					action.TimeStamp,
					'QUEST_COMPLETED',
					GameEventHelper.CreateProvider(
						'QUEST_COMPLETED',
						cardId,
						controllerId,
						entity.Id,
						this.StateFacade,
					),
					true,
					node,
				),
			];
		}
		return null;
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/QuestCompletedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class QuestCompletedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public QuestCompletedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            Parser.ReplayData.GameActions.Action action = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(Parser.ReplayData.GameActions.Action)
                && (action = node.Object as Parser.ReplayData.GameActions.Action).Type == (int)BlockType.TRIGGER
                // BG seems to only need sidequest
                && (action.TriggerKeyword == (int)GameTag.SIDE_QUEST || action.TriggerKeyword == (int)GameTag.QUEST);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var action = node.Object as Parser.ReplayData.GameActions.Action;
            var entity = GameState.CurrentEntities[action.Entity];
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            if (GameState.CurrentEntities[action.Entity].GetTag(GameTag.CARDTYPE) != (int)CardType.ENCHANTMENT)
            {
                var parentAction = (node.Parent.Object as Parser.ReplayData.GameActions.Action);
                //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, null, null);
                entity.SetTag(GameTag.SECRET_HAS_TRIGGERED, 1);
                return new List<GameEventProvider> {
                    GameEventProvider.Create(
                        action.TimeStamp,
                        "QUEST_COMPLETED",
                        GameEvent.CreateProvider(
                            "QUEST_COMPLETED",
                            cardId,
                            controllerId,
                            entity.Id,
                            StateFacade),
                       true,
                       node) 
                };
            }
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
